### PR TITLE
VP-2860: Fix error product quantity in FFC decreases when "Track Inventory" = False

### PR DESCRIPTION
**Pull request by trueboroda**

### CODE REVIEW

File: src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs [hunk 1]
```diff
@@ -136,12 +140,14 @@ protected virtual async Task TryAdjustOrderInventory(ProductInventoryChange[] pr
             var inventoryAdjustments = new HashSet<InventoryInfo>();
             //Load all inventories records for all changes and old order items
             var productIds = productInventoryChanges.Select(x => x.ProductId).Distinct().ToArray();
+            var products = await _itemService.GetByIdsAsync(productIds, ItemResponseGroup.None.ToString());
             var inventoryInfos = await _inventoryService.GetProductsInventoryInfosAsync(productIds);
             foreach (var productInventoryChange in productInventoryChanges)
             {
                 var inventoryInfo = inventoryInfos.Where(x => x.FulfillmentCenterId == (productInventoryChange.FulfillmentCenterId ?? x.FulfillmentCenterId))
```
Comment (t13ka) [inline src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:147]: "inventoryInfos"
possible multiple enumeration

Comment (t13ka) [inline src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:147]: can be simplified, eg:
"var inventoryInfo = inventoryInfos.FirstOrDefault(
                    x => x.FulfillmentCenterId == (productInventoryChange.FulfillmentCenterId ?? x.FulfillmentCenterId)
                         && x.ProductId.EqualsInvariant(productInventoryChange.ProductId));"

**Comment by t13ka** [comment]
Can we add a new unit test for the new conditions in logics?

File: src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs [hunk 2]
```diff
@@ -136,12 +140,14 @@ protected virtual async Task TryAdjustOrderInventory(ProductInventoryChange[] pr
...
                     .FirstOrDefault(x => x.ProductId.EqualsInvariant(productInventoryChange.ProductId));
-                if (inventoryInfo != null)
+                var product = products.FirstOrDefault(x => x.Id.EqualsInvariant(productInventoryChange.ProductId));
+                if (inventoryInfo != null && (product?.TrackInventory ?? false))
```
[the remark this answers is not in the document]
Comment (tatarincev) [reply src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:150]: please look at the  vc-ci recommendation

[the remark this answers is not in the document]
Comment (trueboroda) [reply src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:150]: @tatarincev  Track inventory have a 'bool?' (nullable bool) type.

File: src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs [hunk 3]
```diff
@@ -35,18 +37,20 @@ public class AdjustInventoryOrderChangedEventHandler : IEventHandler<OrderChange
...
+        private readonly IItemService _itemService;
 
         /// <summary>
         /// Constructor.
         /// </summary>
         /// <param name="inventoryService">Inventory service to use for adjusting inventories.</param>
         /// <param name="storeService">Implementation of store service.</param>
         /// <param name="settingsManager">Implementation of settings manager.</param>
-        public AdjustInventoryOrderChangedEventHandler(IInventoryService inventoryService, IStoreService storeService, ISettingsManager settingsManager)
+        public AdjustInventoryOrderChangedEventHandler(IInventoryService inventoryService, IStoreService storeService, ISettingsManager settingsManager, IItemService itemService)
```
Comment (t13ka) [inline src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:48]: missed the documentation tag

Comment (t13ka) [reply src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:48]: <param name="itemService"></param>

File: src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs [hunk 4]
```diff
@@ -136,16 +140,18 @@ protected virtual async Task TryAdjustOrderInventory(ProductInventoryChange[] pr
...
             var productIds = productInventoryChanges.Select(x => x.ProductId).Distinct().ToArray();
-            var inventoryInfos = await _inventoryService.GetProductsInventoryInfosAsync(productIds);
+            var products = await _itemService.GetByIdsAsync(productIds, ItemResponseGroup.None.ToString());
+            var inventoryInfos = (await _inventoryService.GetProductsInventoryInfosAsync(productIds)).ToArray();
             foreach (var productInventoryChange in productInventoryChanges)
             {
-                var inventoryInfo = inventoryInfos.Where(x => x.FulfillmentCenterId == (productInventoryChange.FulfillmentCenterId ?? x.FulfillmentCenterId))
-                    .FirstOrDefault(x => x.ProductId.EqualsInvariant(productInventoryChange.ProductId));
-                if (inventoryInfo != null)
+                var inventoryInfo = inventoryInfos.FirstOrDefault(x => x.FulfillmentCenterId == (productInventoryChange.FulfillmentCenterId ?? x.FulfillmentCenterId)
```
Comment (t13ka) [inline src/VirtoCommerce.OrdersModule.Data/Handlers/AdjustInventoryOrderChangedEventHandler.cs:147]: quite difficult to understand the logic in this expression.
Why we can't just use "
**var inventoryInfo = inventoryInfos.FirstOrDefault(x => x.FulfillmentCenterId == productInventoryChange.FulfillmentCenterId && x.ProductId.EqualsInvariant(productInventoryChange.ProductId));**
"
?

**Pull request merged**